feat: gh-ost integration user doc (#344)
bytebase/bytebase.com: content/docs/en/gh-ost-integration.md

---
title: Use gh-ost in Bytebase
description: This guide shows you how to use gh-ost to migrate your MySQL databases in Bytebase. 
---

> gh-ost is a triggerless online schema migration tool for MySQL.

Reviewer: let's add a link to the gh-ost github project page

All existing online-schema-change tools operate similarly: they create a ghost table in the likeness of your original table, migrate that table while empty, slowly and incrementally copy data from your original table to the ghost table, meanwhile propagating ongoing changes (any INSERT, DELETE, UPDATE applied to your table) to the ghost table. Finally, at the right time, they replace your original table with the ghost table.

> Bytebase uses the power of gh-ost to help users migrate MySQL databases with little downtime.

Reviewer: ```suggestion
Bytebase leverages gh-ost to migrate MySQL table schemas with little downtime.
```

<hint-block type="warning">

> This is currently a preview feature. Data loss is possible.

Reviewer: ```suggestion
This is currently a preview feature.
```

Reviewer: Let's move this to the top

</hint-block>

## Requirements

- MySQL versions 5.7 or greater.
- Enable row-based logging.

> - Bytebase currently only supports migrating on main.

Reviewer: This is not clear

## Limitations

> - Foreign key constraints are not supported.

Reviewer: Should this be put under Requirement as well?

The cutover task atomically renames `yourtablename`, `~yourtablename_{timestamp}_gho` to `~yourtablename_{timestamp}_del`, `yourtablename` respectively to switch the original table and the ghost table.

Click “Approve” to perform the cutover task.

![The issue detail page where the cutover task is waiting approval.](/static/docs/gh-ost-step-3-1.webp)

### Step 4 - Delete `~yourtablename_{timestamp}_del` after migration

After migration, the original table is renamed to `~yourtablename_{timestamp}_del`. Make sure there is no data loss, then manually drop the original table if you wish. You can check the table by clicking “Show Bytebase reserved tables” on the database page.

![A table list where "Show Bytebase reserved tables" button is highlighted](/static/docs/gh-ost-step-4-1.webp)
![A table list which also shows reserved tables](/static/docs/gh-ost-step-4-2.webp)

## Interact with gh-ost

Gh-ost listens on a UNIX socket file.

The UNIX socket file name is `/tmp/gh-ost.{taskID}.{databaseID}.{databaseName}.{tableName}.sock`

> To find the UNIX socket file, we must acquire the database id and task id.

Reviewer: Use "you" instead of "we" across the article.
